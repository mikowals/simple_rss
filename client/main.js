var DAY = 1000 * 60 * 60 * 24;
var articlesOnLoad = 20;
var updateNowFreq = 1 * 60 * 1000;

var intervalProcesses = []; //hold interval process id to start and stop with different functions.
Session.setDefault("loaded", false);
Session.setDefault("importOPML", false);
Session.setDefault( "now", new Date() );
Session.setDefault("articleLimit", articlesOnLoad );

var cleanForXml = function ( string ){
  string = string.replace ( /\"/g, "&quot;");
  string = string.replace ( /\'/g, "&apos;");
  string = string.replace ( /\&/g, "&amp;");
  string = string.replace ( /\>/g, "&gt;");
  
  return  string.replace ( /\</g, "&lt;");
};
                   
var articleSub;

Deps.autorun( function(){
  console.log( Session.get( "articleLimit" ) );
  articleSub = Meteor.subscribe( "feedsWithArticles", + Session.get( "articleLimit" ) );
});

Deps.autorun( function(){
  if ( articleSub && articleSub.ready()) 
    Session.set( "loaded",  true);
});

Meteor.startup( function() {
                          
  intervalProcesses['updateNow'] = intervalProcesses['updateNow'] || Meteor.setInterval( function() {
    Session.set( "now", new Date() );
    },
    updateNowFreq );
  Session.set( "offline", null);

});

Deps.autorun( function(){
    if ( ! Meteor.status().connected && Session.equals( "loaded", true) ) {
      console.log( "Meteor.status().connected = " + Meteor.status().connected );
      Session.set ("offline", "offline" );
    }
    else if ( articleSub && articleSub.ready() ){
      Session.set("offline", null);
    }
});


var timeago = function( some_date ){
  var now = new Date( Session.get( "now" ) );
  var referenceDate = new Date( some_date );

  var timeago = ( now -  referenceDate ) / DAY;

  if (Math.floor(timeago )  >= 2) return Math.floor(timeago ) + " days ago";
  else if (Math.floor(timeago )  >= 1) return Math.floor(timeago ) + " day ago";
  else if (Math.floor(timeago  * 24)  >= 2 ) return Math.floor(timeago * 24) + " hours ago";
  else if (Math.floor(timeago * 24)  >= 1 ) return Math.floor(timeago  * 24) + " hour ago";
  else if (Math.floor(timeago  * 24 * 60) >= 2) return Math.floor(timeago * 24 * 60) + " minutes ago";
  else {
    return "about a minute ago";
  }
};

Handlebars.registerHelper('timeago',  function(some_date){
  return timeago(some_date);
});

Template.feedList.feeds= function () {
  return Feeds.find({}, {sort: {title: 1}});
};

Template.feedList.flash = function(){
  return Session.get("feedListFlash");
};

Template.modalButtons.importOPML = function(){
  return Session.equals("importOPML", true);
};

Template.modalButtons.events({
                             
                             //could modify this to verify feed and populate fields for insertion
                             'submit, click #addFeed': function(e) {
                               e.stopImmediatePropagation();
                               e.preventDefault();
                               var url = $("#feedUrl").val();
                               var regex =/(((f|ht){1}tp|tps:\/\/)[-a-zA-Z0-9@:%_\+.~#?&\/\/=]+)/g;
                               if ( Match.test ( url, String ) && regex.test(url) ){
                                 Meteor.call( 'addFeed', { url: url } );
                                 $("#feedUrl").val("");
                               } else{
                                 alert("RSS feed entered is not a valid url");
                               }
                               return false;        
                             },
                             
                             'click #importToggle': function(){
                             Session.set("importOPML", true);
                             console.log(Session.equals("importOPML", true));
                             },
                             
                             'click #opmlUpload' : function(){
                               Session.set("importOPML", false);
                             
                               var opmlFile = $("#opmlFile")[0].files[0];
                               var fr = new FileReader();
                               fr.readAsText(opmlFile);
                               fr.onloadend = function(evt) {
                                 if (evt.target.readyState === FileReader.DONE) { // DONE == 2
                                   //Meteor.call( 'XML2JSparse', evt.target.result, function ( error, result ){
                                   $(  evt.target.result ).find( 'outline').each( function( ){
                                      var url = $( this ).attr("xmlUrl");
                                      if ( url )
                                        Meteor.call('addFeed', {url: url} );
                                   });
                                 }
                               }
                             },
                          
                             'click #importCancel' : function(){
                             Session.set("importOPML", false);
                             },
                             
                             'click #exportOPML': function(){
                               var exportOPML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
                                               "<opml version=\"1.0\">" +
                                                "<head>" +
                                                 "<title>" + Meteor.user.username + " subscriptions from New-River</title>" +
                                                 "</head>" +
                                               "<body>";
                               Feeds.find().forEach( function( feed ) {
                                 exportOPML += "<outline " +
                                               "text=\"" + cleanForXml( feed.title ) + "\" " +
                                               "title=\"" + cleanForXml( feed.title ) + "\" " +
                                               "type=\"rss\" " +
                                               "xmlUrl=\"" + cleanForXml( feed.url ) + "\"/>";
                               });
                                exportOPML += "</body></opml>";
                                //exportOPML = ( new window.DOMParser() ).parseFromString( exportOPML, "text/xml");
                                //exportOPML = (new XMLSerializer()).serializeToString( exportOPML );
                                var blob = new Blob([exportOPML], {type: "application/xml"});
                                var fname = Meteor.absoluteUrl().split( "//" )[1];
                                saveAs ( blob, fname + ".opml");
                             }
                         
                             });

Template.feedList.events({                                                   
                         'click #removeFeed': function(){
                           Meteor.call( 'removeFeed', Session.get("selected_feed"), function( error ){
                             if ( error === false ) console.error( "could not remove feed"); 
                             else Session.set( "selected_feed", null);
                             });
                         }
                         
                         });
                           
Template.articleList.articles = function() {
  
   return Articles.find( {}, { sort: { date: -1, _id: 1} } );
};

Template.articleList.loaded = function(){
  return Session.equals( "loaded", true );
};

Template.articleList.events({
  'click a, contextmenu a': function( e ){
     if ( ! Session.equals( "handleTap", true)){
       Meteor.call( 'markRead', $( e.currentTarget ).attr('href'));
     } else{
       e.preventDefault();
       e.stopImmediatePropagation();
       return false;
     }
     return true;
  },
  
  'tap a':  function( e ){
    Session.set( "handleTap", true);
    e.preventDefault();
    e.stopImmediatePropagation();
    var dest = $( e.currentTarget ).attr('href');
    Meteor.call( 'markRead', dest, function(){
      window.location.href = dest;
    });
    return false;
  }
}); 

Template.article.subscribed = function(){
  return Feeds.findOne({title: this.source}) !==null;
};

Template.article.subscribeRss = function(){
  if (this.sourceUrl) return this.sourceUrl;

  var feed = Feeds.findOne( {title: this.source});
  return feed && feed.url;
  
};

Template.menubar.loaded = function(){
  return Session.equals( "loaded", true );
};

Template.menubar.offline = function(){
  return Session.get("offline");
}

Template.updated.updated = function(){
  return Session.get( "updated" );
}

Template.feed.selected = function () {
  return Session.equals("selected_feed", this._id) ? "selected" : '';
};

Template.feed.article_count = function () {
  return Articles.find({source: this.title}).count();
};

Template.feed.events({
                     'click': function(){
                     Session.set("selected_feed", this._id);
                     }
                     });


Template.articleList.rendered = function(){

  $('#footer').waypoint( function( evt, dir) {
    if ( dir === 'down' )
      Session.set( "articleLimit", Session.get( "articleLimit" ) + 20 );
    },
    { offset: '110%'}); 
};
