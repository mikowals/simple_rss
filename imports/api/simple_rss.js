import { Mongo } from 'meteor/mongo';
import { Meteor } from 'meteor/meteor';

export const Feeds = new Mongo.Collection ( "feeds" );
export const Articles = new Mongo.Collection ( "articles" );
export const initialArticleLimit = 40;
export const DAY = 1000 * 60 * 60 * 24;
export const keepLimitDate = function(){
  return new Date( new Date().getTime() - ( DAY * daysStoreArticles ));
};
export const daysStoreArticles = 3.0;

if (Meteor.isServer) {
  BrowserPolicy.content.allowConnectOrigin("*.mak-play.com");
  BrowserPolicy.content.allowEval();

  Feeds._ensureIndex( { url: 1 }, {unique: true} );
  Articles._ensureIndex( { link: 1 }, {unique: true, dropDups: true });
  Articles._ensureIndex( { feed_id: 1, date: -1} );

  //Accounts.config({sendVerificationEmail: true});
  Meteor.users.deny({
    update: () => true
  });
}
