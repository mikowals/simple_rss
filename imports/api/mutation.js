import { gql } from '@apollo/client';

export const REMOVE_FEED = gql`
  mutation removeFeed($id: String) {
    removeFeed(id: $id) {
      _id
    }
  }
`;

export const ADD_FEED = gql`
mutation addFeed($_id: String, $url: String!) {
  addFeed(_id: $_id, url: $url) {
    _id
    title
    date
    url
    count
  }
}
`
