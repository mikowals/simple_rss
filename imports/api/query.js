import { gql } from '@apollo/client';

export const ARTICLES_QUERY = gql`
  query($userId: String!) {
    articles(userId: $userId) {
      _id
      source
      date
      title
      link
      summary
      feed_id
    }
  }
`;

export const FEEDS_QUERY = gql`
  query feeds($userId: String!) {
    feeds(userId: $userId) {
        _id
        title
        url
        last_date
        count
    }
  }
`;

export const FEED_COUNT = gql`
  query articlesCountByFeed($id: String) {
    articlesCountByFeed(id: $id) {
      _id
      count
    }
  }
`;

export const FEED_IDS = gql`
  query feedIds($userId: String!) {
    feedIds(userId: $userId)
  }
`

export const USER_QUERY = gql`
  query user($userId: String!) {
    user(userId: $userId) {
      _id
      feeds {
        _id
        title
        url
        last_date
        count
      }
    }
  }
`
