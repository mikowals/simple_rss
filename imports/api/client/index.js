import React from 'react';
import { hydrateRoot } from 'react-dom';
import { BrowserRouter, Route, Routes, Link } from 'react-router-dom';
import { Meteor } from 'meteor/meteor';
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client';
import { BatchHttpLink } from "@apollo/client/link/batch-http";
import { FeedsPageWithContainer } from '/imports/ui/feeds';
import { ArticlesPageWithStream } from '/imports/ui/articles';

export const client = new ApolloClient({
  cache: new InMemoryCache().restore(window.__APOLLO_STATE__),
  link: new BatchHttpLink({
    uri: '/graphql',
    credentials: 'same-origin',
    batchInterval: 10}),
  connectToDevTools: true
});

export const renderRoutes = () => (
    <BrowserRouter>
      <ApolloProvider client={client}>
        <Routes>
          <Route exact path="/feeds" element={<FeedsPageWithContainer />} />
          <Route exact path="/articles" element={<ArticlesPageWithStream />} />
          <Route element={<ArticlesPageWithStream />} />
        </Routes>
      </ApolloProvider>
      <Link to="/feeds" >Feeds</Link>
      <Link to="/articles" >Articles</Link>
    </BrowserRouter>
);


Meteor.startup(() => {
  hydrateRoot(
    document.getElementById('app'),
    renderRoutes()
  )
});
