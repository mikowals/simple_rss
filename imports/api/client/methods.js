import { Meteor } from 'meteor/meteor';
import { Articles, Feeds } from '/imports/api/simple_rss';

var handleError = function( error ){
  if ( error ) console.log( error );
};

Meteor.methods({
  removeFeed: function (feedId) {
    var self = this;
    var userId = self.userId || 'nullUser';
    check(feedId, String);

    var feed = Feeds.findOne( {_id: feedId, subscribers: userId}, {fields:{_id: 1, subscribers: 1, url: 1, hub: 1}});

    Articles.remove({feed_id: feedId});
    Feeds.remove(feedId);
  },

  addFeed: function( doc ){
    var self = this;
    self.unblock();
    check ( doc, { url: String});
    var originalUrl = doc.url;
    var userId = self.userId || 'nullUser';

    // create id so we can insert articles before feed itself is inserted.
    doc._id = Feeds._makeNewID();
    if ( self.isSimulation ) {
      Feeds.insert( doc, handleError );
      return;
    }
  }
});
