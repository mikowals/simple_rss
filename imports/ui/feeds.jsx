import React, { useState, useEffect, memo } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Feeds } from '/imports/api/simple_rss';
import { TimeAgoContainer } from './timeAgo';
import { initialArticleLimit } from '/imports/api/simple_rss';
import { useQuery, useMutation, gql, useLazyQuery } from '@apollo/client';
import { USER_QUERY, FEEDS_QUERY } from '/imports/api/query';
import { ADD_FEED, REMOVE_FEED } from '/imports/api/mutation';
import orderBy from 'lodash.orderby';

export const FeedsPageWithContainer = () => (
  <div id="feed-container">
    <FeedsPage />
  </div>
);

// Spread the feed object (...) to avoid rerendering
const renderFeed = (feed) => <Feed {...feed} key={feed._id} />;

export const FeedsPage = () => {
  const self = this;
  const {loading, error, data} = useQuery(
    FEEDS_QUERY,{
      variables: {userId: "nullUser"},
      fetchPolicy: "cache-and-network"
    }
  );

  if (error) { console.log(error) }
  let feedDiv = <div />;
  if (data) {
    const sortedFeeds = orderBy(data.feeds, [feed => feed.title.toLowerCase()]);
    feedDiv = <div>{sortedFeeds.map(renderFeed)}</div>;
  }
  return (
    <>
      <AddFeed />
      <span className="pad-top-5 col-xs-6 col-md-6"><h4><strong>Feed Title</strong></h4></span>
      <span className="col-xs-2 col-md-2 text-right"><h4><strong>Count</strong></h4></span>
      <span className="col-xs-4 text-right"><h4><strong>Last update</strong></h4></span>
      {feedDiv}
    </>
  );
};

const Feed = memo(({_id, url, title, last_date, count}) => {
  const self = this;
  return <React.Fragment>
           <h5 className="col-xs-7 col-md-7 pull-left">
             <Remove _id={_id} />
             <a  href={url}> {title || ""}</a>
           </h5>
           <h5 className="count col-xs-1 col-md-1 text-right">
             <FeedCount count={count}/>
           </h5>
           <h5 className="lastDate time col-xs-2 col-md-4 text-right pull-right">
             <TimeAgoContainer date={last_date} />
           </h5>
         </React.Fragment>;
});

Feed.propTypes = {
  title: PropTypes.string,
  url: PropTypes.string.isRequired,
  last_date: PropTypes.number,
  _id: PropTypes.string.isRequired,
  count: PropTypes.number.isRequired
};

Feed.displayName = 'Feed';

const FeedCount = ({count}) => {

  return <span>{count}</span>;
};

FeedCount.propTypes = {count: PropTypes.number};

const Remove = memo(({_id}) => {
  const [onDeleteHandler] = useMutation(REMOVE_FEED, {
    update(cache, { data: { removeFeed } }) {
      let articlesToRemove = [];
      cache.modify({
        id: "ROOT_QUERY",
        fields: {
          feeds(existingFeedRefs, { readField }) {
            return existingFeedRefs.filter(
              feedRef => removeFeed._id !== readField('_id', feedRef)
            );
          },
          articles(existingArticleRefs, { readField }) {
            articlesToRemove = existingArticleRefs.filter(
              ref => readField("feed_id", ref) === removeFeed._id
            );
            return existingArticleRefs.filter(
              ref => readField("feed_id", ref) !== removeFeed._id
            );
          }
        },
      });
      cache.evict({ id: cache.identify(removeFeed) });
      articlesToRemove.forEach(ref => cache.evict({id: cache.identify(ref)}));
      cache.gc();
    }
  });
  const handleClick = (e) => {
    onDeleteHandler({
      variables: {id: _id},
      optimisticResponse: {
        __typename: "Mutation",
        removeFeed: {
          __typname: "Feed",
          _id: _id
        }
      }
    })
  };
  return <a onClick={handleClick}>
           <i className="glyphicon glyphicon-remove-circle"></i>
         </a>;
});

Remove.displayName = "Remove";

const AddFeed = memo(() => {
  let [newURL, setNewURL] = useState("");
  let [addHandler] = useMutation( ADD_FEED, {
    update(cache, { data: { addFeed } }) {
      cache.modify({
        id: "ROOT_QUERY",
        fields: {
          feeds(existingRefs, { readField }) {
            let newRef = cache.writeFragment({
              data: addFeed,
              fragment: gql`
                fragment NewFeed on Feed {
                  _id
                  title
                  url
                  last_date
                  count
                }
              `
            });
            if (existingRefs.some(
              ref => readField('url', ref) === addFeed.url
            )) {
              return existingRefs;
            }

            return [...existingRefs, newRef];
          },
        }
      });
    }
  });
  const handleSubmit = (e) => {
    //Stop submit from navigating away from feeds page.
    e.preventDefault();
    e.stopPropagation();
    if (! newURL) {
      alert("URL can not be empty");
      return;
    }

    // Getting the new feed info can be slow.
    // Switch URL to blank so users see action before calling server.
    // Replace with erroring URL so user can check for typos.
    const url = newURL
    const _id = Feeds._makeNewID();
    setNewURL("");
    addHandler({variables: {_id, url},
      optimisticResponse: {
        __typename: "Mutation",
        addFeed: {
          _id: _id,
          title: "adding...",
          url: url,
          last_date: Date.now(),
          count: 0,
          __typename: "Feed"
        }
      }
    });
  };

  const handleInput = (e) => setNewURL(e.target.value);

  return <React.Fragment>
         <form className="form-inline pull-right col-sm-12" onSubmit={handleSubmit}>
           <span className="mt-1 col-xs-0 col-sm-0"/>
           <span className="input-group input-group-sm col-xs-12 col-sm-12 pull-right">
             <input onChange={handleInput} type="url" value={newURL} className="input-sm col-xs-12" placeholder="http://new URL/rss.xml" id="feedUrl" />
             <a id="addFeed" onClick={handleSubmit} onTouchStart={handleSubmit} type="submit" className="input-group-addon btn btn-primary btn-sm pull-right">Subscribe</a>
           </span>
         </form>
         <hr className="col-sm-12"/>
        </React.Fragment>;
});

AddFeed.diplayName = "AddFeed";
