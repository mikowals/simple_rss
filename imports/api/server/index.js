import { ApolloServer, gql } from 'apollo-server-express';
import { WebApp } from 'meteor/webapp';
import { Feeds, Articles } from '/imports/api/simple_rss';
import React from 'react';
import { renderToStaticMarkup, renderToNodeStream, renderToString } from 'react-dom/server';
import { ApolloClient, InMemoryCache, ApolloProvider, createHttpLink} from '@apollo/client';
import { renderToStringWithData } from "@apollo/client/react/ssr";
import fetch from 'cross-fetch';
import { ApolloApp } from '/imports/ui/app';
import { FeedsPage } from '/imports/ui/feeds';
import { ArticlesPage } from '/imports/ui/articles'
import { ServerStyleSheet } from "styled-components";
import { Meteor } from 'meteor/meteor';
import { onPageLoad } from 'meteor/server-render';
import { resolvers } from '/imports/api/server/resolvers';
import { typeDefs } from '/imports/api/server/typeDefs';
//import { WebApp } from 'meteor/webapp'
//import { makeExecutableSchema } from 'graphql-tools';

const server = new ApolloServer({typeDefs, resolvers});

server.applyMiddleware({
  app: WebApp.connectHandlers,
  path: '/graphql'
})

BrowserPolicy.content.allowOriginForAll("https://fonts.googleapis.com");
BrowserPolicy.content.allowOriginForAll("http://cdn.jsdelivr.net");
BrowserPolicy.content.allowOriginForAll("http://fonts.gstatic.com");
// We are doing this work-around because Playground sets headers and WebApp also sets headers
// Resulting into a conflict and a server side exception of "Headers already sent"
WebApp.connectHandlers.use('/graphql', (req, res) => {
  if (req.method === 'GET') {
    res.end()
  }
});

// Wrap ArticlesPage here so that we can inject it into div id='stream'.
const SSRPage = ({client, location}) => (
  <ApolloProvider client={client}>
    {location === "/feeds" ? <FeedsPage /> : <ArticlesPage />}
  </ApolloProvider>
);

function Html({ content, state }) {
  return (
    <html>
      <body>
        <div id="app">
        <div id="stream" dangerouslySetInnerHTML={{ __html: content }} />
        </div>
        <script dangerouslySetInnerHTML={{
          __html: `window.__APOLLO_STATE__=${JSON.stringify(state).replace(/</g, '\\u003c')};`,
        }} />

      </body>
    </html>
  );
}

function AppWithCache({ content, state, location }) {
  return (
    <>
    {<div id={location === "/feeds" ? "feed-container" : "stream"}
      dangerouslySetInnerHTML={{ __html: content }} />}
    <script dangerouslySetInnerHTML={{
      __html: `window.__APOLLO_STATE__=${JSON.stringify(state).replace(/</g, '\\u003c')};`,
    }} />
    </>
  );
}

// this is faster than server-render 'onPageLoad' but lacks css and js to continue updates.
WebApp.connectHandlers.use('/static', (req, res, next) => {
  const client = new ApolloClient({
    ssrMode: true,
    // Remember that this is the interface the SSR server will use to connect to the
    // API server, so we need to ensure it isn't firewalled, etc
    link: createHttpLink({
      uri: 'http://localhost:3000/graphql',
      credentials: 'same-origin',
      fetch
    }),
    cache: new InMemoryCache(),
  });
  renderToStringWithData(SSRPage({client: client})).then((content) => {
    const initialState = client.extract();
    const html = <Html content={content} state={initialState} />;
    res.writeHead(
      200,
      {'Content-Type': 'text/html'}
    );
    renderToNodeStream(html).pipe(res)
  })
});

onPageLoad(async sink => {
  const client = new ApolloClient({
    ssrMode: true,
    // Remember that this is the interface the SSR server will use to connect to the
    // API server, so we need to ensure it isn't firewalled, etc
    link: createHttpLink({
      uri: 'http://localhost:3000/graphql',
      credentials: 'same-origin',
      fetch
    }),
    cache: new InMemoryCache(),
  });

  const sheet = new ServerStyleSheet();
  const location = sink.request.url.pathname;
  const content = await renderToStringWithData(SSRPage({client, location}));

  const initialState = client.extract();
  const appJSX = <AppWithCache content={content} state={initialState} location={location} />;

  const htmlStream = renderToNodeStream(appJSX);
  sink.renderIntoElementById("app", htmlStream);

});
