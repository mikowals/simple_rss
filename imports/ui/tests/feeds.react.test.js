import React from 'react';
import {cleanup, MockedProvider} from '@apollo/client/testing';
import { InMemoryCache } from '@apollo/client';
import renderer, {act} from 'react-test-renderer';
//import { fireEvent } from '@testing-library/react';
import { Simulate } from 'react-dom/test-utils';
import {FeedsPage, AddFeed} from '../feeds';
import { FEEDS_QUERY } from '../../api/query';
import { REMOVE_FEED, ADD_FEED } from '../../api/mutation';
import "babel-polyfill";
import wait from 'waait';

const feed1 = {
  _id: '1',
  title: 'a publisher',
  url: 'http://blah.blah.com',
  date: (new Date()).getTime(),
  count: 3
};

const feed2 = {
  _id: '2',
  title: "another publisher",
  url: "http://more.blah.com",
  date: (new Date()).getTime() - 2000,
  count: 4
}

// For insertion tests the __typename fields allow apollo to generate cache ids.
const feed3 = {
  _id: '3',
  title: 'some guy',
  url: 'http://blah_blah.com',
  date: (new Date()).getTime(),
  count: 5,
  __typename: "Feed",
};

// This object is to be inserted.
// Its _id of '4' matches the return of _makeNewID mock in /.meteorMocks/index.js.
const feed4 = {
  _id: '4',
  title: "a woman",
  url: "http://again_with_the_blah.com",
  date: (new Date()).getTime() - 2000,
  count: 6,
  __typename: "Feed",
}

let cache = new InMemoryCache();

beforeEach(() => {
  cache = new InMemoryCache();
});

it('renders without error', async () => {
  const mocks = [
    {
      request: {
        query: FEEDS_QUERY,
        variables: { userId: 'nullUser'}
      },
      result: {
        data: {
          feeds: [feed1, feed2]
        }
      },
    },
  ];
  const component = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <FeedsPage />
    </MockedProvider>,
  );

  await act(wait);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('can remove a feed', async () => {
  let mutationCalled = false;
  const mocks = [
    {
      request: {
        query: FEEDS_QUERY,
        variables: { userId: 'nullUser'}
      },
      result: {
        data: {
          feeds: [feed1, feed2]
        }
      },
    },{
      request: {
        query: REMOVE_FEED,
        variables: { id: '1'}
      },
      result: () => {
        mutationCalled = true;
        return {
          data: {
            removeFeed: {_id: "1"}
          }
        }
      },
    },{
      request: {
        query: FEEDS_QUERY,
        variables: { userId: 'nullUser'}
      },
      result: {
        data: {
          feeds: [feed2]
        }
      },
    },
  ];
  const component = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <FeedsPage />
    </MockedProvider>,
  );

  await act(wait);
  const listItems = component.root.findAllByType('i');
  expect(listItems).toHaveLength(2)

  act(listItems[0].parent.props.onClick);
  await act(wait);
  expect(mutationCalled).toBe(true);
  const listItems2 = component.root.findAllByType('i');
  expect(listItems2).toHaveLength(1)
  await act(wait);
});

it('can add a feed', async () => {
  let mutationCalled = false;
  const mocks = [
    {
      request: {
        query: FEEDS_QUERY,
        variables: { userId: 'nullUser'}
      },
      result: {
        data: {
          feeds: [feed3]
        }
      },
    },{
      // _id: '4' matches the variables generated by mocked _makeNewID.
      request: {
        query: ADD_FEED,
        variables: { _id: '4', url:"http://another.com"}
      },
      result: () => {
        mutationCalled = true;
        return {
          data: {
            __typename: "mutation",
            addFeed: feed4,
          }
        }
      },
    },{
      request: {
        query: FEEDS_QUERY,
        variables: { userId: 'nullUser'}
      },
      result: {
        data: {
          feeds: [feed3, feed4]
        }
      },
    },
  ];
  const component = renderer.create(
    <MockedProvider mocks={mocks} addTypename={true}>
      <FeedsPage />
    </MockedProvider>,
  );

  await act(wait);
  const listItems = component.root.findAllByType('li');
  expect(listItems).toHaveLength(1)

  const input = component.root.findByType('input');
  act(() => input.props.onChange({target: {value: "http://another.com"}}));
  expect(input.props.value).toBe("http://another.com");
  const button = component.root.findByType('form').findByType('a');
  act(() => {
    button.props.onClick({
      preventDefault: () => {},
      stopPropagation: () => {}
    })
  });
  //Simulate.click(button);
  expect(input.props.value).toBe("");
  await act(wait);

  expect(mutationCalled).toBe(true);
  const listItemsAfter = component.root.findAllByType('li');
  expect(listItemsAfter).toHaveLength(2)
});

// This needs tests for failing cases.
// Remove id that doesn't exist.
// Add duplicate url.
// Submitting form before input has value.
