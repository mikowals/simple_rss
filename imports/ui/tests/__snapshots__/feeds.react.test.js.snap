// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without error 1`] = `
Array [
  <form
    className="form-inline pull-right col-sm-12"
    onSubmit={[Function]}
  >
    <span
      className="mt-1 col-xs-0 col-sm-0"
    />
    <span
      className="input-group input-group-sm col-xs-12 col-sm-12 pull-right"
    >
      <input
        className="input-sm col-xs-12"
        id="feedUrl"
        onChange={[Function]}
        placeholder="http://new URL/rss.xml"
        type="url"
        value=""
      />
      <a
        className="input-group-addon btn btn-primary btn-sm pull-right"
        id="addFeed"
        onClick={[Function]}
        onTouchStart={[Function]}
        type="submit"
      >
        Subscribe
      </a>
    </span>
  </form>,
  <hr
    className="col-sm-12"
  />,
  <span
    className="pad-top-5 col-xs-6 col-md-6"
  >
    <h4>
      <strong>
        Feed Title
      </strong>
    </h4>
  </span>,
  <span
    className="col-xs-2 col-md-2 text-right"
  >
    <h4>
      <strong>
        Count
      </strong>
    </h4>
  </span>,
  <span
    className="col-xs-4 text-right"
  >
    <h4>
      <strong>
        Last update
      </strong>
    </h4>
  </span>,
  <div>
    <li>
      <h5
        className="col-xs-7 col-md-7 pull-left"
      >
        <a
          id="_id + _remove"
          onClick={[Function]}
        >
          <i
            className="glyphicon glyphicon-remove-circle"
          />
        </a>
        <a
          href="http://blah.blah.com"
        >
           
          a publisher
        </a>
      </h5>
      <h5
        className="count col-xs-1 col-md-1 text-right"
      >
        <span>
          3
        </span>
      </h5>
      <h5
        className="lastDate time col-xs-2 col-md-4 text-right pull-right"
      >
        <span>
          less than a minute ago
        </span>
      </h5>
    </li>
    <li>
      <h5
        className="col-xs-7 col-md-7 pull-left"
      >
        <a
          id="_id + _remove"
          onClick={[Function]}
        >
          <i
            className="glyphicon glyphicon-remove-circle"
          />
        </a>
        <a
          href="http://more.blah.com"
        >
           
          another publisher
        </a>
      </h5>
      <h5
        className="count col-xs-1 col-md-1 text-right"
      >
        <span>
          4
        </span>
      </h5>
      <h5
        className="lastDate time col-xs-2 col-md-4 text-right pull-right"
      >
        <span>
          less than a minute ago
        </span>
      </h5>
    </li>
  </div>,
]
`;
